lib/Processor.php

```diff
@@ -147,52 +143,54 @@ private function ready() {
         }
     }
 
-    public function connect() {
+    public function connect(): Promise {
         \assert(!$this->deferreds && !$this->socket, self::class."::connect() must not be called twice");
 
-        $this->deferreds[] = $deferred = new Deferred;
-        \Amp\Socket\connect($this->config->resolvedHost)->onResolve(function ($error, $socket) use ($deferred) {
-            if ($this->connectionState === self::CLOSED) {
```

Comment:
we get $this->connectionState === self::CLOSED here when the user calls closeSocket() before connection is established. This must be safe here.